hbase-server/src/main/java/org/apache/hadoop/hbase/master/normalizer/SimpleRegionNormalizer.java

```diff
@@ -101,11 +106,21 @@ public void onConfigurationChange(Configuration conf) {
     setConf(conf);
   }
 
-  private static int parseMinRegionCount(final Configuration conf) {
-    final int parsedValue = conf.getInt(MIN_REGION_COUNT_KEY, DEFAULT_MIN_REGION_COUNT);
+  private static int parseMergeMinRegionCount(final Configuration conf) {
+    String parsedStringValue = conf.get(MERGE_MIN_REGION_COUNT_KEY);
```

Comment:
I _believe_ you can register the old key via the [deprecation API](https://hadoop.apache.org/docs/r2.10.1/api/org/apache/hadoop/conf/Configuration.html#addDeprecation(java.lang.String,%20java.lang.String,%20java.lang.String)) and only ever access the new key from code. I believe this has the benefit of only logging the warning once, instead of every time the implementation happens to do a lookup.

Reply:
The deprecation API is good, but has some confusion as I said in the jira. When we call conf.addDeprecation(deprecatedConf, newConf), and only set newConf,  conf.get(deprecatedConf) always get value even it not set.

Reply:
The key point is how to recognize that the user has used the old configuration and log warnings once.

Reply:
And I found after conf.addDeprecation(deprecatedConf, newConf), the conf.get(deprecatedConf) can not get value even we configured in conf file. So if the configuration is not modified before upgrade to HBase-2.5, then the 
configuration will be use default value. I don't know if i express cleared...

Reply:
The example of the use of conf deprecation from block cache that you mention on Jira is, in my opinion, an example that should not be followed.

 > And I found after conf.addDeprecation(deprecatedConf, newConf), the conf.get(deprecatedConf) can not get value even we configured in conf file. So if the configuration is not modified before upgrade to HBase-2.5, then the configuration will be use default value.

Okay, this is bad.

Reply:
I think we need to keep the old config in hbase-defaults.xml and add the new config there as well.

```
modified   hbase-common/src/main/resources/hbase-default.xml
@@ -41,6 +41,8 @@ possible configurations would overwhelm and obscure the important.
 <configuration>
   <!--Configs you will likely change are listed here at the top of the file.
   -->
+  <property><name>foo</name><value>42</value><description>deprecated; see 'bar'</description></property>
+  <property><name>bar</name><value>42</value></property>
   <property >
     <name>hbase.tmp.dir</name>
     <value>${java.io.tmpdir}/hbase-${user.name}</value>
---
|  Welcome to JShell -- Version 11.0.10
|  For an introduction type: /help intro

jshell> import org.apache.hadoop.conf.Configuration

jshell> import org.apache.hadoop.hbase.HBaseConfiguration

jshell> Configuration c = HBaseConfiguration.create()
c ==> Configuration: core-default.xml, core-site.xml, h ... efault.xml, hbase-site.xml

jshell> c.get("foo")
$4 ==> "42"

jshell> c.get("bar")
$5 ==> "42"

jshell> Configuration.addDeprecation("foo", "bar")

jshell> c.get("foo")
2021-05-03T14:03:00,038 INFO  [main] Configuration.deprecation: foo is deprecated. Instead, use bar
$7 ==> "42"

jshell> c.get("bar")
$8 ==> "42"

jshell> c.set("foo", "99")

jshell> c.get("foo")
$10 ==> "99"

jshell> c.get("bar")
$11 ==> "99"
```

Reply:
Your example is special because you call the c.set() method. But we all define the configuration in hbase-site.xml. If your hbase-site.xml define: <property><name>foo</name><value>99</value></property> and do not call c.set("foo", "99"), the value of c.get("bar") and c.get("foo") is "42". If you call Configuration.addDeprecation("foo", "bar") first, then call Configuration.addDeprecation("foo", "bar"), the value of c.get("bar") and c.get("foo") is 99.